src/ripple/shamap/README.md

```diff
@@ -26,164 +45,322 @@ There are two different ways of building and using a SHAMap:
 
-It is cheaper to make an immutable snapshot of a SHAMap than to make a mutable
-snapshot.  If the SHAMap snapshot is mutable then any of the nodes that might
-be modified must be copied before they are placed in the mutable map.
+The nodes of a SHAMap have their own copy of a sequence number.  If the SHAMap
+is not immutable, meaning it can change, then all of its nodes must have the
+same sequence number as the SHAMap itself.  This enforces an invariant that none
+of the nodes are shared with other SHAMaps.
 
+When a SHAMap gets a copy of a node from elsewhere, and must have a private
+copy, it clones the node and sets the nodes sequence number equal to the SHAMap
+sequence number.
 
-## SHAMap Thread Safety ##
+When a SHAMap decides that it is safe to share a node of its own, it sets the
+node's sequence number to 0 (a SHAMap never has a sequence number of 0). This
+is done for every node in the trie when `SHAMap::walkSubTree` is executed.
 
-*This description is obsolete and needs to be rewritten.*
+Note that other objects in rippled also have sequence numbers (e.g. ledgers).
+The SHAMap and node sequence numbers should not be confused with these other
+sequence numbers (no relation).
 
-SHAMaps can be thread safe, depending on how they are used.  The SHAMap
-uses a SyncUnorderedMap for its storage.  The SyncUnorderedMap has three
-thread-safe methods:
-
- * size(),
- * canonicalize(), and
- * retrieve()
+## SHAMap Creation ##
 
-As long as the SHAMap uses only those three interfaces to its storage
-(the mTNByID variable [which stands for Tree Node by ID]) the SHAMap is
-thread safe.
+A SHAMap is usually not created from vacuum.  Once an initial SHAMap is
+constructed, later SHAMaps are usually created by calling snapShot(bool
+isMutable) on the original SHAMap().  The returned SHAMap has the expected
+characteristics (mutable or immutable) based on the passed in flag.
 
+It is cheaper to make an immutable snapshot of a SHAMap than to make a mutable
+snapshot.  If the SHAMap snapshot is mutable then any of the nodes that might be
+modified must be copied before they are placed in the mutable map.
+
+A new SHAMap is created with each new ledger round.  Transactions not executed
+in the previous ledger populate the SHAMap for the new ledger.
+
+## Storing SHAMap data in the database ##
+
+When consensus is reached, the ledger is closed.  As part of this process, the
+SHAMap is stored to the database by calling `SHAMap::flushDirty` with parameters
+which specify which nodes are to be written.
+
+Both `unshare()` and `flushDirty` walk the SHAMap by calling
+`SHAMap::walkSubTree`.  As `unshare()` walks the trie, nodes are not written to
+the database, and as `flushDirty` walks the trie nodes are written to the
+database. `walkSubTree` visits every node in the trie. This process must ensure
+that each node is only owned by this trie, and so performs a COW operation
+(`unshare()`) as it walks each node (from the root down).  This is done by the
```

Comment:
> so performs a COW operation (`unshare()`) as it walks each node (from the root down).

This is a bit confusing to me, because it seems to be saying that `walkSubTree` calls `unshare`, but when I read the code, I don't see `walkSubTree` calling `unshare()` anywhere. Am I missing something in the code, or am I misunderstanding this sentence?